### PR TITLE
Cancellation support for lexing and parsing ops in language service features

### PR DESCRIPTION
**Pull request by T-Gro**
This adds support for cancellation during lexing+parsing.
This does not affect the code path being taken by standalone compiler (that ones goes via `ParseOneInputLexbuf` at ParseAndCheckInputs.fs ;; which was not modified).

The cancellation token flows from higher-level operations which were already `async{}` before, this extracts the `let! ct = Async.CancellationToken` from it.

(cc @vzarytovskii  This does mean that if operations inside Compiler.Service will be redone from async to cancellableTask, the token extraction must be also adjusted)

I tested this in VS, the operation was being cancelled under quick typing (before I adjusted the exception handlers to correctly handle the cancelled exn, it was all going into VS activity log and/or showing up an alert dialog)

### CODE REVIEW

**Comment by vzarytovskii** [comment]
> @vzarytovskii This does mean that if operations inside Compiler.Service will be redone from async to cancellableTask, the token extraction must be also adjusted

Yeah, currently, if you call getDefaultCancellationToken, it will return one which was passed to the topmost cancellableTask. Thinking about it. It should be renamed. To just "cancellationToken".

File: src/Compiler/Service/FSharpCheckerResults.fs [hunk 1]
```diff
@@ -2357,12 +2357,25 @@ module internal ParseAndCheckFile =
         let tokenizer =
             LexFilter.LexFilter(indentationSyntaxStatus, options.CompilingFSharpCore, Lexer.token lexargs true, lexbuf, false)
 
-        (fun _ -> tokenizer.GetToken())
+        if ct.CanBeCanceled then
+            (fun _ ->
+                ct.ThrowIfCancellationRequested()
```
Comment (baronfel) [inline src/Compiler/Service/FSharpCheckerResults.fs:2362]: How big of an impact on perf does checking for cancellation at every token have? That seems _very_ granular - should we instead check every N tokens/end of line/etc?

Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: I was thinking about this as well, but in the end this is what the check does:
https://referencesource.microsoft.com/#mscorlib/system/threading/CancellationTokenSource.cs,716cb764064ecfeb

And I would need to replace it with mutable counter management (increase, compare) and doing e.g. a modulo N operation.

Comment (0101) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: Maybe `let` and `member`?

Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: in .fsi files both can be fully missing I guess?

File: src/Compiler/Service/FSharpCheckerResults.fs [hunk 2]
```diff
@@ -2357,12 +2357,25 @@ module internal ParseAndCheckFile =
...
+                tokenizer.GetToken())
+        else
+            (fun _ -> tokenizer.GetToken())
 
     let createLexbuf langVersion strictIndentation sourceText =
         UnicodeLexing.SourceTextAsLexbuf(true, LanguageVersion(langVersion), strictIndentation, sourceText)
 
-    let matchBraces (sourceText: ISourceText, fileName, options: FSharpParsingOptions, userOpName: string, suggestNamesForErrors: bool) =
+    let matchBraces
```
Comment (0101) [inline src/Compiler/Service/FSharpCheckerResults.fs:2370]: So what about making `matchBraces` and `parseFile` `Cancellable` instead of manually passing in the token?

File: src/Compiler/Service/FSharpCheckerResults.fs (hunk 1 above)
Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: The cancellation check is really a simple op, you might actually force me for proving that :-D

File: src/Compiler/Service/FSharpCheckerResults.fs (hunk 2 above)
Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2370]: It's internal and I covered all usages => I guess it does not matter?

File: src/Compiler/Service/FSharpCheckerResults.fs (hunk 1 above)
Comment (0101) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: `member` would be in fsi file as well, so I suppose also add `val` ? But maybe that's unnecessarily granular, so perhaps `type`?

Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: Now it definitely feels overcomplicated just to balance a simple token check.
Both 'type' as well as 'member' can be fully missing in a file.

Comment (auduchinok) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: Would it be cheap to check a token line here? We could probably check if cancellation needed when line changes, or even every few lines.

File: src/Compiler/Service/FSharpCheckerResults.fs (hunk 2 above)
Comment (0101) [reply src/Compiler/Service/FSharpCheckerResults.fs:2370]: I just thought it would be more consistent since we use `Cancellable` for this purpose elsewhere.

File: src/Compiler/Service/FSharpCheckerResults.fs (hunk 1 above)
Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: I did a loop from 0 to 10_000, which is only doing a basic multiplication by two in every operation - a single instruction, so this should really highlight the cost differences between cancellation token check or anything more complex.

NaiveCheckNoneToken  - check each iteration, but it is a CancellationToken.None
NaiveCheckRealToken  - check each iteration, but this token can be cancelled
CheckFewHardcodedOps  - this does basic branching similar to how this would need to branch on type of token
CheckRealTokenEveryFewOps  - this does need to keep a separate counter, and do a 'modulo` calculation on every iteration 

|                    Method | every |      Mean |     Error |    StdDev | Allocated |
|-------------------------- |------ |----------:|----------:|----------:|----------:|
|       NaiveCheckNoneToken |     ? |  6.216 us | 0.1164 us | 0.1032 us |         - |
|       NaiveCheckRealToken |     ? |  7.581 us | 0.0774 us | 0.0604 us |         - |
|      CheckFewHardcodedOps |     ? |  6.635 us | 0.0793 us | 0.0662 us |         - |
| CheckRealTokenEveryFewOps |    10 | 13.275 us | 0.0790 us | 0.0700 us |         - |
| CheckRealTokenEveryFewOps |   100 | 13.718 us | 0.1547 us | 0.1372 us |         - |
| CheckRealTokenEveryFewOps |   500 | 13.076 us | 0.2246 us | 0.1991 us |         - |
| CheckRealTokenEveryFewOps |  1000 | 12.932 us | 0.1433 us | 0.1340 us |         - |



=> `ct.ThrowIfCancellationRequested()` is really as lightweight as it can get, I do not think we should bother about it at all.

Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: NOTE: This was really a 1-op situation repeated 10_000 times, hence this really shows just the overhead differences in a very artificial scenario.
Even though real lexing is super fast, there are still a few more things going on (deciding based on characters, indexing into tables, allocating a token type,...) - the ThrowIfCancellationRequested() really should not be a worry in the case of not being cancelled.


Things do get a cost hit when the operation is being cancelled, that is given - all of this PR is under the assumption that things in the IDE toolchain would deal with cancellation later anyway, so the cost for throwing exception will always be exist somewhere, and with this PR it will just happen sooner.

Comment (baronfel) [reply src/Compiler/Service/FSharpCheckerResults.fs:2362]: Thanks for investigating this - I'm now convinced that the naieve approach is best :)

**Review by 0101 — APPROVED** [review]

File: src/Compiler/Service/FSharpCheckerResults.fs [hunk 3]
```diff
@@ -2357,12 +2357,25 @@ module internal ParseAndCheckFile =
         let tokenizer =
             LexFilter.LexFilter(indentationSyntaxStatus, options.CompilingFSharpCore, Lexer.token lexargs true, lexbuf, false)
 
-        (fun _ -> tokenizer.GetToken())
+        if ct.CanBeCanceled then
```
Comment (vzarytovskii) [inline src/Compiler/Service/FSharpCheckerResults.fs:2360]: Is the check necessary? All it does, checks CTS for null, right? And the method already does that. In case if we will be calling that lambda with the check in it, JIT will likely optimize the call to be a nop if CTS is default/non-cancellable.

Comment (vzarytovskii) [reply src/Compiler/Service/FSharpCheckerResults.fs:2360]: Though it will ikely be inlined either way.

Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2360]: The difference is that the code added by me is invoked just once, and later on the lambda is created.
If we are e.g. passed a CancellationToken.None , then just the plain lambda is returned and no subsequent checking happens at all, never.

TLDR:
The difference is between making the check just once before creating the func, or inside the func each time.

**Review by abonie — APPROVED** [review]

File: src/Compiler/Service/FSharpCheckerResults.fs (hunk 3 above)
Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2464]: I tested it - you are right under the circumstance of everything being inline from the place where token is being created.
(which might be well inside the VisualStudio / Rider /.. dlls).
In the inlined case, I can indeed see the "nop"s.

But when it is not inlined, the asm contains the full check.

https://sharplab.io/#v2:DYLgZgzgNALiCGEC2AfA9gBwKYDsAEAygJ4QxZIDcAsAFCa6EllIB0AKgBYBOW8AJgEscAc1piawLDDwAKAFRDgQrHICUeGBwEQAkhACyRAGIBXHAGNZ5uAGF4FrMGDwYAtDjZoA1rnUBeWjwgvElpJBMYeAAjSTwAdw4XAFEANywuPD88AEZA4LA0DIE/AAYNNBySqrw+NDzg4Ot2bjQ4nTA7BycXNxwAJSwARxMsUiw+GVV6hvjEmFT0vAAeAFpZ5LSMuTwAJmmG/eDaULwcVqNCgfhgWX9pk4APTI0tXQNjM0sZTvNHZ1d3J4fDgWAA5dxYKY0BoPaZwiRSU7nS68YA7W6Ze6IiBoExcX7PHBYOJ4H5/HqA7y4Ai4/FYSZY6RPLKabR6QymCyyHF437sKk4KEwoA=

=> Since the callstack up to this point is large and not controlled by us, I would rather not assume inlining happens and leave the code as is
